grr/server/grr_response_server/gui/webauth.py

```diff
@@ -51,6 +53,57 @@ def RedirectBase(self):
     """Return a redirect to the main GRR page."""
     return werkzeug_utils.redirect(config.CONFIG["AdminUI.url"])
 
+class IAPWebAuthManager(BaseWebAuthManager):
+  """Auth Manager for Google IAP.
+
+  This extension pulls the x-goog-iap-jwt-assertion header and generates
+  a new user for that header via the 'sub' claim. Authorization is now 
+  delegated to IAP.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super(IAPWebAuthManager, self).__init__(*args, **kwargs)
+
+    if config.CONFIG["AdminUI.cloud_project_id"] is None or \
+       config.CONFIG["AdminUI.cloud_backend_service_id"] is None:
+      raise RuntimeError("The necessary Cloud IAP configuration options are"
+                         "not set. Please set your AdminUI.cloud_project_id"
+                         "or AdminUI.cloud_backend_service_id keys.")
+
+    self.cloud_project_id = config.CONFIG["AdminUI.cloud_project_id"]
+    self.backend_service_id = config.CONFIG["AdminUI.cloud_backend_service_id"]
+    self.iap_header = "x-goog-iap-jwt-assertion"
+
+  def SecurityCheck(self, func, request, *args, **kwargs):
+    """Wrapping function."""
+    request.user = u""
+
+    if self.iap_header in request.headers:
+      jwt = request.headers.get(self.iap_header)
+      user_id, user_email, error_str = validate_iap.ValidateIapJwtFromComputeEngine(jwt, self.cloud_project_id, self.backend_service_id)
+    else:
+      return werkzeug_wrappers.Response("Unauthorized", status=401)
+
+    if error_str:
+      # Log error
+      logging.error("IAPWebAuthManager failed with: %s", error_str)
+
+      # Return failure if IAP is not decoded correctly
+      return werkzeug_wrappers.Response("Unauthorized", status=401)
+    else:
+      # Generate a new user if not created, else authenticate current
+      try:
+        grr_user = config_updater_util.GetUserSummary(user_id)
+      except config_updater_util.UserNotFoundError:
+        grr_user = None
+
+      if grr_user:
```

Comment:
I think just doing `request.user = user_id` should be enough here. GRR will create the user automatically, you don't have to do it in the webauth manager.

Reply:
I seeded each user with a long random password so a user who is logged in can't switch users to another user who has logged in via IAP, thus obfuscating logging. Do you think this will be an issue if I resolve this comment?

Reply:
Ok, I think I understand the scenario. Quick question - do you plan to have 2 AdminUI's running, one for basic http-based auth and one for IAP? Because at the moment GRR AdminUI can have only one WebAuth manager set, meaning that a single AdminUI instance would be able to run with *either* IAP or HTTP Basic auth.

Reply:
Ah, that makes sense. We have two admin UIs loaded in our implementation, but they are ostensibly going to deploy the same Docker image. I think what you point out here means that we wouldn't need to generate a new seeded password each time. I'll do some testing of new account creation to validate, and push your suggestion.